web3/utils/events.py

```diff
@@ -238,3 +253,168 @@ def remove_trailing_from_seq(seq, remove_value=None):
+        self._toBlock = value
+
+    @property
+    def address(self):
+        return self._address
+
+    @address.setter
+    def address(self, value):
+        self._address = value
+
+    @property
+    @to_tuple
+    def indexed_args(self):
+        return filter(
+            is_indexed,
+            map(lambda x: self.args[x], self._ordered_arg_names))
+
+    @property
+    @to_tuple
+    def data_args(self):
+        return filter(
+            not_indexed,
+            map(lambda x: self.args[x], self._ordered_arg_names))
+
+    @property
+    def topics(self):
+        arg_topics = tuple(arg.match_values for arg in self.indexed_args)
+        return normalize_topic_list(cons(to_hex(self.event_topic), arg_topics))
+
+    @property
+    def data_argument_values(self):
+        if self.data_args is not None:
+            return tuple(arg.match_values for arg in self.data_args)
+        else:
+            return (None,)
+
+    @property
+    def filter_params(self):
+        params = {
+            "topics": self.topics,
+            "fromBlock": self.fromBlock,
+            "toBlock": self.toBlock,
+            "address": self.address
+        }
+        return valfilter(lambda x: x is not None, params)
+
+    def deploy(self, web3):
+        if web3 is None:
+            raise ValueError("Missing required argument 'web3'")
+
+        log_filter = web3.eth.filter(self.filter_params)
+        log_filter.filter_params = self.filter_params
+        log_filter.set_data_filters(self.data_argument_values)
+        log_filter.builder = self
+        if self.formatter is not None:
+            log_filter.log_entry_formatter = self.formatter
+        return log_filter
+
+    def _initial_event_topic(self):
+        if self.event_abi['anonymous'] is False:
+            return event_abi_to_log_topic(self.event_abi)
+        else:
+            return list()
+
+
+@to_dict
+def _build_argument_filters_from_event_abi(event_abi):
+    for item in event_abi['inputs']:
+        if item['indexed'] is True:
+            argument_filter_class = TopicArgumentFilter
+        else:
+            argument_filter_class = DataArgumentFilter
+        key = item['name']
+        value = argument_filter_class(arg_type=item['type'])
+
+        yield key, value
+
+
+array_to_tuple = apply_formatter_if(is_list_like, tuple)
+
+
+class BaseArgumentFilter(ABC):
+    _match_values = None
+
+    def __init__(self, arg_type):
+        self.arg_type = arg_type
+
+    def match_single(self, value):
+        self._match_values = self._normalize_match_values((value,))
+
+    def match_any(self, *values):
+        self._match_values = self._normalize_match_values(values)
+
+    @to_tuple
+    def _normalize_match_values(self, match_values):
+        for value in match_values:
+            yield array_to_tuple(value)
+
+    @property
+    @abstractmethod
+    def match_values(self):
+        raise NotImplementedError("ArgumentFilter classes must implement match_values method")
```

Comment:
I think I'd like to move away from using `NotImplementedError` here and just put `pass` statements since the `abstractmethod` decorator raises an exception for us on instantiation if this method has not been implemented.